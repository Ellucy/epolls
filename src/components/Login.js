import { useNavigate, useLocation } from "react-router-dom";
import { useState, useEffect } from "react";
import { connect } from 'react-redux';

function Login({ users }) {
    const navigate = useNavigate();
    const { pathname } = useLocation();
    const [selectedUserId, setSelectedUserId] = useState("none");
    const [selectedUser, setSelectedUser] = useState({});

    useEffect(() => {
        if (users[selectedUserId]) {
            setSelectedUser(users[selectedUserId]);
            return
        }
        setSelectedUser({});

    }, [selectedUserId, users]);

    return (
        <div className="App padding10">
            <div className="signin-container">
                <div className="login-header">Employee Polls</div>
                <select
                    value={selectedUserId}
                    className="login-padding10"
                    onChange={(e) => {
                        setSelectedUserId(e.target.value)
                    }}
                >
                    <option value="none" disabled>
                        Select user...
                    </option>
                    {Object.values(users).map((u) => (
                        <option
                            key={u.id}
                            value={u.id}
                        >
                            {u.name}
                        </option>
                    ))}
                </select>
                <div>
                    {selectedUser?.avatarURL && (
                        <img src={selectedUser.avatarURL} alt={`Avatar of ${selectedUser.id}`} className="big-avatar" />
                    )}
                </div>
                {selectedUserId !== "none" ?
                    (
                        <button
                            className="submit-button"
                            onClick={() => {
                                navigate(
                                    pathname,
                                    {
                                        state: { userId: selectedUserId },
                                        replace: true
                                    }
                                );
                            }}
                        >
                            Submit
                        </button>
                    ) : null}

            </div>
        </div >
    );
}

const mapStateToProps = ({ users }) => ({
    users,
});

export default connect(mapStateToProps)(Login);