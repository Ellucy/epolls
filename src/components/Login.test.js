const mockedUsedNavigate = jest.fn();
const mockedUsedLocation = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
    useLocation: () => mockedUsedLocation,

}));

import { render, fireEvent } from '@testing-library/react';
import { Login } from './Login';

const usersObj = {
    sarahedo: {
        id: 'sarahedo',
        password: 'password123',
        name: 'Sarah Edo',
        avatarURL: "https://tylermcginnis.com/would-you-rather/sarah.jpg",
        answers: {
            "8xf0y6ziyjabvozdd253nd": 'optionOne',
            "6ni6ok3ym7mf1p33lnez": 'optionOne',
            "am8ehyc8byjqgar0jgpub9": 'optionTwo',
            "loxhs1bqm25b708cmbf3g": 'optionTwo'
        },
        questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
    },
    tylermcginnis: {
        id: 'tylermcginnis',
        password: 'abc321',
        name: 'Tyler McGinnis',
        avatarURL: "https://tylermcginnis.com/would-you-rather/tyler.jpg",
        answers: {
            "vthrdm985a262al8qx3do": 'optionOne',
            "xj352vofupe1dqz9emx13r": 'optionTwo',
        },
        questions: ['loxhs1bqm25b708cmbf3g', 'vthrdm985a262al8qx3do'],
    },
};

describe('Greetings', () => {

    it('matches the snapshot when empty user object passed', () => {
        const component = render(<Login users={{}} />);
        expect(component).toMatchSnapshot();
    });

    it('matches the snapshot when user object passed', () => {
        const component = render(<Login users={usersObj} />);
        expect(component).toMatchSnapshot();
    });

    it('checks that log in button is not present initially', () => {
        const component = render(<Login users={usersObj} />);
        expect(component.queryByTestId('submit-button')).not.toBeInTheDocument();
    });

    it('checks that log in button is present after user has been selected', () => {
        const component = render(<Login users={usersObj} />);
        fireEvent.change(component.getByTestId('select-user'), { target: { value: 'sarahedo' } })
        expect(component.queryByTestId('submit-button')).toBeInTheDocument();
    });

});
