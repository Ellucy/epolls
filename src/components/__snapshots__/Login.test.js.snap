// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Greetings matches the snapshot when empty user object passed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App padding10"
      >
        <div
          class="signin-container"
        >
          <div
            class="login-header"
          >
            Employee Polls
          </div>
          <select
            class="login-padding10"
            data-testid="select-user"
          >
            <option
              disabled=""
              value="none"
            >
              Select user...
            </option>
          </select>
          <div />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App padding10"
    >
      <div
        class="signin-container"
      >
        <div
          class="login-header"
        >
          Employee Polls
        </div>
        <select
          class="login-padding10"
          data-testid="select-user"
        >
          <option
            disabled=""
            value="none"
          >
            Select user...
          </option>
        </select>
        <div />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Greetings matches the snapshot when user object passed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App padding10"
      >
        <div
          class="signin-container"
        >
          <div
            class="login-header"
          >
            Employee Polls
          </div>
          <select
            class="login-padding10"
            data-testid="select-user"
          >
            <option
              disabled=""
              value="none"
            >
              Select user...
            </option>
            <option
              value="sarahedo"
            >
              Sarah Edo
            </option>
            <option
              value="tylermcginnis"
            >
              Tyler McGinnis
            </option>
          </select>
          <div />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App padding10"
    >
      <div
        class="signin-container"
      >
        <div
          class="login-header"
        >
          Employee Polls
        </div>
        <select
          class="login-padding10"
          data-testid="select-user"
        >
          <option
            disabled=""
            value="none"
          >
            Select user...
          </option>
          <option
            value="sarahedo"
          >
            Sarah Edo
          </option>
          <option
            value="tylermcginnis"
          >
            Tyler McGinnis
          </option>
        </select>
        <div />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
